### PR TITLE
Support for UQ with conformal prediction

### LINKED ISSUE #848: Add support for UQ with conformal prediction
Conformal prediction is well established for IID data, which is the majority of our current use cases. This should be a straightforward way to add principled UQ to the deterministic emulators (when there is enough data available for  the calibration set).

### PR DESCRIPTION
**Pull request by sgreenbury**
Closes #848.
Contributes towards #589.
(merge after #917)

This PR:
- Updates the base emulator API to enable the specification of validation data (to be used here as calibration data) (#905)
- Adds a base class wrapper for emulators to provide conformal prediction (analogous to `Ensemble`)
- Adds a `ConformalMLP` class (analogous to `EnsembleMLP`). This supports both constant width and quantile regression derived UQ.
- The quantile regression version uses an MLP with a quantile training loss to fit the upper and lower quantiles and then uses a constant per-target correction based on performance on the calibration training data
- Adds `ConformalMLP` to registry and re-exports (not adding as default emulator)
- Updates API to support passing `n_samples` - this is to enable this to be controlled at lower levels of the API from the high-level `AutoEmulate`.
- Adds `output_to_tensors` method on `ConversionMixin` to avoid repeated mapping of outputs to tensors
- Fixes indentation of docstrings in MLP subclasses

Question:
- Currently the two methods for deriving the intervals are part of the same base class with a parameter passed as a keyword argument - this is similar to the way the kernel fn is passed for GPs. Would it be preferable to have two classes? Or perhaps a general emulator subclass factory extending the one currently that only supports GPs.

Next steps:
- [x] Tests for passing validation data
- [x] Check UQ performance of `ConformalMLP` (e.g. on projectile and other simulators)
  - [x] For projectile "split" constant-width intervals does not provide good UQ while "quantile" is an improvement but less good than e.g. a GP.

### CODE REVIEW

**Review by radka-j — COMMENTED** [review]
This is looking great, I am just writing down some immediate thoughts I had after a quick look before I have a chance to go through it in more detail.

1:  Method naming/structure: There are different ways to handle heteroscedacity with conformal methods. As here, one can use some kind of model that predicts quantiles (this can be quantile regression but could also be any other model like random forest or a NN). Alternatively, people have focused on predicting a scaling factor $\sigma_i$ that is also a function of data that is then used to scale the intervals (again, there are many ways that $\sigma_i$ can be computed. Therefore it feels that we should explicitly call this `ConformalizedQuantileMLP` or something like that (see e.g., [paper](https://proceedings.neurips.cc/paper_files/paper/2019/file/5103c3584b063c431bd1268e9b5e76fb-Paper.pdf)).

2: What are the pros/cons of using an MLP vs just a standard quantised regression here?

3: I think it would be better to write the `ConformalEmulator` more along the lines of `TransformedEmulator` i.e., a wrapper that can take any deterministic model (rather than having a `ConformalMLP` specifically. It feels that all the conformal components are model agnostic. Is the issue here that we need a name for the emulator to pass to `AutoEmulate()`?

**Review by radka-j — APPROVED** [review]
This looks great! All my comments are really just about naming/docstrings.

**Comment by sgreenbury** [comment]
> I don't think this PR closes #905 since we still don't let the user specify the test data set (and use all `x` as train data).

Ah I see - #905 is distinct from updates to the `Emulator` API required in #589. I'll update the top-level comment as contributes towards #589. Are the API changes in #905 specifically towards the `AutoEmulate` class?

**Comment by radka-j** [comment]
I was thinking about our discussion yesterday about evaluating conformal intervals using MSLL. I only had a brief look but I don't think we can treat the conformal intervals as a distribution - we don't know if it is normal but also don't have a reason to assume it is uniform. We only have the 2 quantiles we compute (and the mean).

File: autoemulate/emulators/conformal.py [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+    supports_uq = True
+
+    def __init__(
+        self,
+        emulator: Emulator,
+        alpha: float = 0.95,
+        device: DeviceLike | None = None,
+        calibration_ratio: float = 0.2,
+        n_samples: int = 1000,
+        method: Literal["split", "quantile"] = "split",
```
[the remark this answers is not in the document]
Comment (sgreenbury) [reply autoemulate/emulators/conformal.py]: Yes good point - updated in 183dc76d6f534f

**Comment by EdwinB12** [comment]
Can I check where we are with this PR?

**Comment by radka-j** [comment]
@EdwinB12 waiting for me to do a second review

**Comment by radka-j** [comment]
Closing as CP is now moving to a new package.
